### PR TITLE
#14345 Cupertino Alert Dialog

### PR DESCRIPTION
**Pull request by ekbiker**
Ref issue #14345.
Added the ability to display action button items in vertical list when there are more than two buttons.  The list is scrollable.

### CODE REVIEW

File: examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart
Comment (xster) [inline examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart]: I'd change this text a bit so the actions don't mismatch the title topic

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (xster) [inline packages/flutter/lib/src/cupertino/dialog.dart]: _kEdgePadding

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (xster) [inline packages/flutter/lib/src/cupertino/dialog.dart]: nit: s/Vertical/Vertically/

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (xster) [inline packages/flutter/lib/src/cupertino/dialog.dart]: nit: title on next line too.

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (xster) [inline packages/flutter/lib/src/cupertino/dialog.dart]: Complete sentence comments, e.g. `// Add padding between the sections.`

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (xster) [inline packages/flutter/lib/src/cupertino/dialog.dart]: nit: children on next line

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (xster) [inline packages/flutter/lib/src/cupertino/dialog.dart]: Wrap field references with []. Then it becomes linked in dart docs / API website / IDEs etc. e.g. `// If [title] is missing` etc...

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (xster) [inline packages/flutter/lib/src/cupertino/dialog.dart]: nit: flex on next line too

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (xster) [inline packages/flutter/lib/src/cupertino/dialog.dart]: I'd keep the word divider or something in the name, since it's not painting the actions but the action divider

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (xster) [inline packages/flutter/lib/src/cupertino/dialog.dart]: reduce indent of this block by 2

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (xster) [inline packages/flutter/lib/src/cupertino/dialog.dart]: this works :) 
Might be easier with a ListView with children of interweaved buttons and Containers of 1 height and color too

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (xster) [inline packages/flutter/lib/src/cupertino/dialog.dart]: Ending new line

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (xster) [inline packages/flutter/lib/src/cupertino/dialog.dart]: `// Skip the last one so there's no divider at the very top of the list.`

**Review by xster — COMMENTED** [review]
Awesome! Thanks for your contribution.

Please make sure packages/flutter/test/cupertino/dialog_test.dart passes as well. You can test it by running `flutter test test/cupertino/dialog_test.dart` in the packages/flutter directory.

Please also add a new test in the file for the new behavior.

**Comment by ekbiker** [comment]
@xster, will do.

**Comment by xster** [comment]
Please let me know when this PR is ready for review again :)

**Comment by ekbiker** [comment]
@xster, please review.

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (xster) [inline packages/flutter/lib/src/cupertino/dialog.dart]: Does this work? I suspect it might crash.

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (xster) [inline packages/flutter/lib/src/cupertino/dialog.dart]: ditto

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (xster) [inline packages/flutter/lib/src/cupertino/dialog.dart]: This is essentially saying, CustomPaint instances might be created all the time while creating the widget tree. When a new one is created, does the existing RenderCustomPaint need to call paint again. Since this painter always paints the same way, you can return false here.

**Review by xster — APPROVED** [review]
Sorry I forgot to get back to this. It looks good minus a few more minor notes. Once you're ready, I'll merge it for you.

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (ekbiker) [reply packages/flutter/lib/src/cupertino/dialog.dart]: Got it.

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (ekbiker) [reply packages/flutter/lib/src/cupertino/dialog.dart]: Yes, it works.  See above comment.

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (ekbiker) [reply packages/flutter/lib/src/cupertino/dialog.dart]: Yes, it works.  The method returning null is private and the code that uses it does null check.  I'd rather not return null but Dart doesn't have optional types like Swift or Kotlin.  I'll comment it in the code to avoid others from questioning.

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (ekbiker) [reply packages/flutter/lib/src/cupertino/dialog.dart]: will do.

**Comment by ekbiker** [comment]
@xster, please review.

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (xster) [inline packages/flutter/lib/src/cupertino/dialog.dart]: Ah, I see what you're saying. So you're checking null against the class's construction which will never return null. The widget's build method will be invoked separately by the framework. 

I'd encourage adding a test to exercise this path and the action section's path.

**Review by xster — CHANGES_REQUESTED** [review]

**Comment by ekbiker** [comment]
@xster, I see what you mean by not return null now.  Thanks for all your pointers.  Will update and add tests.

**Comment by ekbiker** [comment]
@xster, please review.

**Review by xster — APPROVED** [review]

File: examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart
Comment (ekbiker) [reply examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart]: @xster, the guidelines don't exhaust all possible examples; the UIAlertController API certainly allows null title/message and null actions.  However, I'm happy to remove these example.

**Comment by ekbiker** [comment]
@xster, updated test and the demo.  Please review, thanks for all your suggestions.

File: packages/flutter/test/cupertino/dialog_test.dart
Comment (xster) [inline packages/flutter/test/cupertino/dialog_test.dart]: just expect(find.widgetWithText(...), findsNothing)

File: packages/flutter/test/cupertino/dialog_test.dart
Comment (xster) [reply packages/flutter/test/cupertino/dialog_test.dart]: What I meant is 

flutter_02

The first button and the second button doesn't have the same height. 

Please check in a test that there is no sizing discrepancy.

**Comment by xster** [comment]
You'd have to make sure the analyzer passes https://travis-ci.org/flutter/flutter/jobs/352367964

You can check locally with `flutter analyze`

File: examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart
Comment (Hixie) [inline examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart]: Please put the space at the end of the first line, not the second. Also, put a newline before the first. As in:

```dart
                 title: const Text(
                   'Allow "Maps" to access your location while '
                   'you use the app?',
                 ),
```
￼...though once you do that, maybe it'll just fit on one line anyway.

File: examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart
Comment (Hixie) [inline examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart]: ditto

Comment (Hixie) [reply examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart]: also, subsequent lines should be vertically aligned with the first

File: examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart
Comment (Hixie) [inline examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart]: ditto

File: examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart
Comment (Hixie) [inline examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart]: why is Four using `"` but the others `'`?

File: examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart
Comment (Hixie) [inline examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart]: for clarity, maybe factor out the desert dialog, and just have both examples call that code. That way they both have interesting dialogs (the numbers aren't super interesting compared to deserts!).

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/dialog.dart]: please make this a TODO, as in `// TODO(xster): This color...`

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/dialog.dart]: those parens are extraneous here

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/dialog.dart]: assert `actions != null`

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/dialog.dart]: In the docs for the constructor just here, add a paragraph that says "The [actions] must not be null."

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/dialog.dart]: strip "note that"

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/dialog.dart]: please don't put a TODO in the api docs.
Just put them on their own line, prefixed by `//`.

But in this case, you can just remove the TODO. Either we make this change now, or we never make it. There's no benefit to waiting.

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/dialog.dart]: why flex 3?

Comment (Hixie) [reply packages/flutter/lib/src/cupertino/dialog.dart]: generally, returning a Flexible is an anti-pattern, since it means the widget can only be used in a Flex. It would be better to move the Flexible into the place where you use this widget.

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/dialog.dart]: These docs should use `//` not `///` since they are not "camera ready".

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/dialog.dart]: assert children != null, since you already guarantee that above

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/dialog.dart]: the children != null part here is redundant

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/dialog.dart]: do you fix this?

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/dialog.dart]: spaces around operators (`index == 0`)

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/dialog.dart]: as before, returning an Expanded is generally an anti-pattern. prefer to move that to the callsite.

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/dialog.dart]: 1 is the default and can be removed.

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/dialog.dart]: same

File: packages/flutter/lib/src/cupertino/picker.dart
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/picker.dart]: this indentation is broken

File: examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart
Comment (ekbiker) [reply examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart]: doesn't continuations indent another two spaces on the second line from the first line?

File: examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart
Comment (ekbiker) [reply examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart]: I'll update accordingly.

File: examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart
Comment (ekbiker) [reply examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart]: :)

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (ekbiker) [reply packages/flutter/lib/src/cupertino/dialog.dart]: I didn't add this comment.

File: examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart
Comment (ekbiker) [reply examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart]: fixed.

File: examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart
Comment (ekbiker) [reply examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart]: fixed

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (ekbiker) [reply packages/flutter/lib/src/cupertino/dialog.dart]: fixed

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (ekbiker) [reply packages/flutter/lib/src/cupertino/dialog.dart]: k

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (ekbiker) [reply packages/flutter/lib/src/cupertino/dialog.dart]: added

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (ekbiker) [reply packages/flutter/lib/src/cupertino/dialog.dart]: fixed

Comment (ekbiker) [reply packages/flutter/lib/src/cupertino/dialog.dart]: I assume you aren't going to make this change. Correct?

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (ekbiker) [reply packages/flutter/lib/src/cupertino/dialog.dart]: What do you mean by Camera Ready?

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (ekbiker) [reply packages/flutter/lib/src/cupertino/dialog.dart]: removed

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (ekbiker) [reply packages/flutter/lib/src/cupertino/dialog.dart]: added

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (ekbiker) [reply packages/flutter/lib/src/cupertino/dialog.dart]: I didn't touch that.

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (ekbiker) [reply packages/flutter/lib/src/cupertino/dialog.dart]: fixed

File: packages/flutter/lib/src/cupertino/picker.dart
Comment (ekbiker) [reply packages/flutter/lib/src/cupertino/picker.dart]: fixed

File: examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart
Comment (Hixie) [reply examples/flutter_gallery/lib/demo/cupertino/cupertino_dialog_demo.dart]: There's various ways to do it, but we try to optimize for readability, so that argues for keeping the strings indented the same amount. See also https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo

File: packages/flutter/lib/src/cupertino/dialog.dart
Comment (Hixie) [reply packages/flutter/lib/src/cupertino/dialog.dart]: We should leave the TODOs then

**Comment by Hixie** [comment]
@ekbiker  Looks like you didn't upload your updated branch.

**Comment by Hixie** [comment]
@ekbiker Please let us know if you be able to continue working on this. If you don't have an opportunity to do so that's totally fine, we will in due course fix the remaining issues and land the patch. Thanks for your contributions so far!

**Comment by ekbiker** [comment]
@Hixie, I apologize for holding this up, I got really busy.  I was so close, sorry I don't have time to finish it. I'll upload what I have.

**Comment by gspencergoog** [comment]
Closing this in favor of https://github.com/flutter/flutter/pull/17676

**Pull request closed without merging**
